### ISSUE TITLE
Review `TODO`, `FIXME`, and Similar Code Comments

### ISSUE THREAD

**Issue by bryanculver** [nautobot/nautobot#2556]
### Proposed Changes
Review the code reminders, todos, fixmes, etc. to identify if they are cruft that needs open issues or aligned to 2.0 breaking changes.

### Justification
Many comments aren't grouped or included in the scoping of housekeeping or future changes for 2.0 and future releases. They should be assignable to address in the backlog or incorporated into an existing story.

**Comment by jathanism** [comment]
# Nautobot #2556 TODO/FIXMEs

# Misc

## `docker/Dockerfile`

It’s not clear why we would want to use `nautobot-server init` here, unless this is indicating that the config that we currently have in `development/nautobot_config.py` was not generated and that it should be?

```
    # TODO Use nautobot init to generate the same config for all use cases
    COPY development/nautobot_config.py /opt/nautobot/nautobot_config.py
```

## `mkdocs.yml`

This appears to be able to be removed as this TODO may no longer be relevant.

```
    # TODO: replace when switching domain
    # site_url: "https://docs.nautobot.com/projects/core/en/stable/"
```

## `pyproject.toml`

There is a rather large list of disabled rules for PyLint that we should revisit.

```
    # TODO: re-enable and fix these as time permits
    # unused-import is already covered by flake8
```

## Various

`**nautobot.utilities.testing.views.APIViewTestCase**`

```
    # TODO: Standardize to APIViewTestCase (needs create & update tests)
```

- Self-explanatory in that create/update generic tests are needed?
- Or is there some clarification required here on what this means?

`**nautobot.utilities.testing.api.TestCase.model_to_dict()**`

```
    # FIXME(jathan): The writable serializer for `status` takes the
    # status `name` (str) and not the `pk` (int). Do not validate this
    # field right now, since we are asserting that it does create correctly.
    #
    # The test code for `utilities.testing.views.TestCase.model_to_dict()`
    # needs to be enhanced to use the actual API serializers when `api=True`
```

- As indicated, the test code in `nautobot.utilities.testing.view.TestCase.model_to_dict()` should know how to handle `SlugRelatedField` types.
- In fact, this should be using the DRF serializers and not the `model_to_dict()` method.

# Core

## `graphql``**/utils.py::get_filtering_args_from_filterset**`

```
    # TODO(jathan): Once we upgrade to Graphene 3.x we can remove this, but we...
```

This should be a v2 TODO at this point. Where we will either upgrade to the latest `graphene-django` that supports Graphene 3.x, or we switch to GraphWrap.

## `tests/test_graphql.py`

`**GraphQLQueryTest.test_query_interfaces_connected_endpoint**`

```
    # TODO: it would be nice to have connections to circuit terminations to test!
```

May be fixed in our fixture factory work?

## `tests/test_views.py`

`**NavRestrictedUI.setUp**`

```
    # TODO: not easy to introspect from the nav menu struct, so hard-code it here for now
```

How can we improve introspection here? Is it worth it?

## `views/generic.py`

`**BulkEditView.post**`

```
    # TODO: Find a better way to accomplish this
```

This is branching logic to extract either `device` or `device_type` from the `request.GET` query parameters, because it is contextually needed for the form. It’s quite possible that this will be obviated as we move toward `NautobotViewSet` and DRF serializers as forms in v2.

`**ComponentCreateView**`

```
    # TODO: Replace with BulkCreateView
```

Another that may very well be obviated by moving to `NautobotViewSet`.

## `views/mixins.py`

`**ObjectEditViewMixin**`

Each of the following methods are overloading methods to work with forms vs. serializers. We should revisit as we move towards serializers as forms.

- `perform_create` → `#` `*TODO: this conflicts with DRF's CreateModelMixin.perform_create(self, serializer) API*`
- `perform_update` → `#` `*TODO: this conflicts with DRF's UpdateModelMixin.perform_update(self, serializer) API*`

`**ObjectBulkUpdateViewMixin**`

We should revisit as we move towards serializers as forms.

- `perform_bulk_update` → `#` `*TODO: this conflicts with BulkUpdateModelMixin.perform_bulk_update(self, objects, update_data, partial)*`

# DCIM

## `forms.py`

`**FrontPortCreateForm**`

```
    # TODO: Merge with FrontPortTemplateCreateForm to remove duplicate logic
```

If we decide to coalesce `FooComponentTemplate` models AND do away with forms in exchange for serializers, this goes away.

## `tests/test_views.py`

`**CableTestCase**` ****

```
    # TODO: Change base class to PrimaryObjectViewTestCase
```

-  Blocked by lack of common creation view for cables (termination A must be initialized)

`**CableTestCase.setUpTestData.form_data**`

```
    # FIXME(John): Revisit this as it is likely an actual bug allowing the terminations to be changed after creation.
```

- This needs to be addressed at the model validation not in the form.

`**DeviceTypeTestCase**` ****

```
    # TODO: Change base class to PrimaryObjectViewTestCase
```

-  Blocked by absence of bulk import view for DeviceTypes

`**InterfaceConnectionsTestCase.test_list_objects_filtered**` ****

```
    # TODO: it'd make test failures more readable if we strip the page headers/footers from the content
```

-  Make `extract_page_body()` also remove the header/footer of the document?

# Extras

## `api/serializers.py`

`**ExportTemplateSerializer**` ****

```
    # TODO: export-templates don't support custom-fields, is this omission intentional?
```

- Export templates should be extended to support custom fields.

`**SecretsGroupSerializer**` ****

```
    # TODO: it would be **awesome** if we could create/update SecretsGroupAssociations
    # alongside creating/updating the base SecretsGroup, but since this is a ManyToManyField with
    # a `through` table, that appears very non-trivial to implement. For now we have this as a
    # read-only field; to create/update SecretsGroupAssociations you must make separate calls to the
    # api/extras/secrets-group-associations/ REST endpoint as appropriate.
```

## `api/views.py`

`**DynamicGroupViewSet.members**`

```
    # FIXME(jathan): Figure out how to do dynamic `responses` serializer based on the `content_type`
    # of the DynamicGroup? May not be possible or even desirable to have a "dynamic schema".
    # @extend_schema(methods=["get"], responses={200: member_response})
```

- This is difficult in that we would need a way to provide the appropriate schema based on the content-type and I don’t think there is a correct fix here other than saying it could be a `list(object)` kind of thing and leave it at that.

`JobViewSet`

~~Both of these issues below appear to be the same, in that the serializer classes are requiring instances to be passed in. So we should rule out if this is necessary, and if it isn’t, don’t. Otherwise, remove the comments.~~

These were removed.

`**list**`

```
    job_model.job_class()  # TODO: why do we need to instantiate the job_class?
```

- This is being used to created a `jobs_list` that gets passed to `JobClassSerializer`

`**retrieve_deprecated**`

```
    job = job_model.job_class()  # TODO: why do we need to instantiate the job_class?
```

- This may very well go away with v2
- In this case this `Job` instance is passed to `JobClassDetailSerializer`

## `constants.py`

`JOB_MAX_NAME_LENGTH`

```
    # TODO(Glenn): this should really be a more general-purpose constant, like NAME_MAX_LENGTH
```

- If so, is this something we would use in other places? What are the references?

## `datasources/git.py`

`**import_config_context**`

```
    # TODO: check context_data against a schema of some sort?
```

- `ConfigContext` import should validate incoming `context_data` against a `ConfigContextSchema`
- The comment is misleading. It is about validating that the `context_data` (the raw contents of a given YAML file) is structured in the way that the Git import expects (includes a `_metadata` key, with appropriate structure beneath that key). The latter part is about associating a ConfigContextSchema record to the newly created/updated ConfigContext record.

`update_git_export_templates`

- This same issue is also found above in `import_config_context()`, so these should be addressed together.

```
    # FIXME: Normally ObjectChange records are automatically generated every time we save an object,
    # regardless of whether any fields were actually modified.
    # Because a single GitRepository may manage dozens of records, this would result in a lot of noise
    # every time a repository gets resynced.
    # To reduce that noise until the base issue is fixed, we need to explicitly detect object changes:
```

- It’s not clear what the “base issue” here is? That `ObjectChange` objects are generated every time we save and not only when a change occurs?

`import_local_config_context`

```
    # TODO: come up with a design that accounts for non-unique names, as well as un-named Devices.
```

- This is because `Device.name` is not guaranteed globally unique
- The correct solution is to make `Device.name` globally unique in a future release.
    - Further un-named devices should not be possible. Sorry.
    - Object in inventory without hostnames should still have hostnames.

## `forms.py`

`RelationshipForm.save`

```
    # TODO add support for owner when a CR is created in the UI
```

- Relationship doesn’t have an `owner`?
- And why is this overloaded? So that we can have the `commit=True` kwarg?

## `jobs.py`

`**BaseJob.class_path**`

```
     TODO(Glenn): it'd be nice if this were derived more automatically instead of needing this logic
```

- Context being probably factoring this into some helper logic, assuming this needs to be reused?

`**_get_job_source_paths**`

```
    # TODO: when a Git repo is deleted or its slug is changed, we update the local filesystem
    # (see extras/signals.py, extras/models/datasources.py), but as noted above, there may be multiple filesystems
    # involved, so not all local clones of deleted Git repositories may have been deleted yet.
    # For now, if we encounter a "leftover" Git repo here, we delete it now.
```

- This follows the theme of similar TODO items for filesystem-based sync for server/workers nodes keeping files in sync “somehow”

## `management/__init__.py`

`**populate_status_choices**`

```
    # TODO: why can't/shouldn't we pass `apps` through to create_custom_statuses? We get failures if we do, but why?
```

- `create_custom_statuses` takes an `apps` argument but it defaults to `global_apps`
- Passing in `apps` appears to break it for some reason, but we don’t know why. So let’s figure that out.

## `models/customfields.py`

`**CustomFieldModel.**``custom_field_data`

```
    # TODO(John): remove this entirely when the cf property is enhanced
```

**
- What exactly needs to be enhanced here?
- Let’s clarify with John A.

`**CustomFieldChoice.delete**`

```
    # TODO: should this be self.field instead?
```

- `protected_objects` passed to `ProtectedError` is sending `self` (The `CustomFieldChoice` instance), but should it be `self.field` (the associated `CustomField`) instead?
- This is a check for this choice being used as a default value
- There are two instances of this call in the `delete()` method

```
    # TODO should this be model.objects.filter(...) instead?
```

- Similar to the above for `ProtectedError` but in the case of a custom field value being set to this choice vs. a default value
- There are two instances of this call, one each for `TYPE_SELECT` and `TYPE_MULTISELECT`

## `models/datasources.py`

`**GitRepository.save.on_commit_callback**`

```
    # TODO: In a distributed Nautobot deployment, each Django instance and/or worker instance may
    # have its own clone of this repository on its own local filesystem; we need some way to ensure
    # that all such clones are renamed.
    # For now we just rename the one that we have locally and rely on other methods
    # (notably get_jobs()) to clean up other clones as they're encountered.
```

- This is is something that needs to be clarified on “how” to keep server and workers in sync when it comes to local copies of files like Git clones on the filesystem for each system.

## `models/groups.py`

`**DynamicGroup**`

- `exclude_filter_fields`

```
    # FIXME(jathan): As one example, `DeviceFilterSet.q` filter searches in `comments`. The issue
    # really being that this field renders as a textarea and it's not cute in the UI. Might be able
    # to dynamically change the widget if we decide we do want to support this field.
```

    - This may be solved already by https://github.com/nautobot/nautobot/pull/2348

- `_map_filter_fields`

```
    # FIXME(jathan); Figure out how we can do this autoamtically from the FilterSet so we
    # don't have to munge it here.
```

    - This may be solved already by https://github.com/nautobot/nautobot/pull/2348

- `get_initial`

```
    # FIXME(jathan): Yes, this is "something", but there is discrepancy between explicitly declared
    # fields on `DeviceFilterForm` (for example) vs. the `DeviceFilterSet` filters. For example
    # `Device.name` becomes a `MultiValueCharFilter` that emits a `MultiValueCharField` which
    # expects a list of strings as input. The inverse is not true. It's easier to munge this
    # dictionary when we go to send it to the form, than it is to dynamically coerce the form field
    # types coming and going... For now.
```

    - Already fixed by https://github.com/nautobot/nautobot/pull/2348

- `generate_filter_form`

```
    # FIXME(jathan): Account for field_order in the newly generated class.
```

    - This entire method can likely go away in exchange for generating a filter form from https://github.com/nautobot/nautobot/pull/2348 helper code?

## `plugins`

`**__init__.py**`

```
    # TODO(jathan): This is fine for now, but as we expand the functionality
    # of plugins, we'll need to consider something like pydantic or attrs.
```

- Each base plugin config item is validated against a type, but it stops there.
- Do we want more robust plugin config validation under “Plugin API”?

`**utils.py**`

- `load_plugin`

```
    # TODO(jathan): We won't be able to support advanced app-ordering concerns
    # and if the time comes that we do, this will have to be rethought.
```

    - Advanced app ordering is a future concern, so we might just want to backlog this under “Plugin API”

## `querysets.py`

`**DynamicGroupQuerySet**`

```
    # FIXME(jathan): Ideally replace this iteration with a reversible Q object
```

and

```
    # FIXME(jathan): Try to get the number of queries down. This will scale poorly.
    # TODO(jathan): 1 query
```

and

```
    # TODO(jathan): 3 queries per DynamicGroup instance
```

and

```
    # TODO(jathan): 1 query
```

- This one is the mofo that is affecting Fiserv at scale for 1,400+ dynamic groups in the context of Golden Config scopes
- We need to codify the `generate_query_` pattern to assert reversible `Q` objects and go from there.

## `signals.py`

`**git_repository_pre_delete**`

```
    # TODO(Glenn): In a distributed Nautobot deployment, each Django instance and/or worker instance may have its own clone
    # of this repository; we need some way to ensure that all such clones are deleted.
    # For now we just delete the one that we have locally and rely on other methods (notably get_jobs())
    # to clean up other clones as they're encountered.
```

- Another one that follows the theme of similar TODO items for filesystem-based sync for server/workers nodes keeping files in sync “somehow”
- It occurred to me that a secondary “cleanup” queue may be required here. Big architecture opportunity here.

`**refresh_job_models**`

```
    # TODO(Glenn): eventually this should be inverted so that get_jobs() relies on the database models...
```

- This idea has merit

```
    # TODO(Glenn): catch DB error in case where multiple Jobs have the same grouping + name
```

- IMO It should not be possible for multiple jobs to have the same grouping + name

## `tables.py`

`**WEBHOOK_CONTENT_TYPES**`

```
    # TODO: Webhook content_types in table order_by
```

- This seems pretty straightforward.

`**JobTable**`

```
    # TODO(Glenn): pk = ToggleColumn()
```

- Why isn’t this already the case?

## `tests`

`**test_api.py**`

- `ImageAttachmentTest` - `#` `*TODO(Glenn): Standardize to APIViewTestCase (needs create & update tests)*`
    - See Misc > Various > `APIViewTestCase` above

`**test_context_managers.py**`

- `WebRequestContextTestCase.test_change_webhook_enqueued`

```
    # TODO(john): come back to this with a way to actually do it without a running worker
    # The celery inspection API expects to be able to communicate with at least 1 running
    # worker and there does not appear to be an easy way to look into the queues directly.
```

    - Seems like the correct solution may to use mocking here?

`**test_datasources.py**`

- `GitTest.populate_repo`  - `#` `*TODO: populate Jobs as well?*`
    - This seems to just be populating test `jobs` published by a Git repo?

`**test_models.py**`

- `JobModelTest.test_latest_result` - `#` `*TODO(Glenn): create some JobResults and test that this works correctly for them as well.*`
    - Possibly this will be addressed by fixture factory work?

`**test_plugins.py**`

- `LoadPluginTest.test_load_plugin`

```
    # FIXME(jathan): We're expecting a PluginNotFound to be raised, but
    # unittest doesn't appear to let that happen and we only see the
    # original ModuleNotFoundError, so this will have to do for now.
```

    - In fact at `PluginNotFound` error is never raised because the condition for it never seems to actually return `True` (`if getattr(err, "name") == plugin_name`)
    - We should just either make it always raise a `PluginNotFound` or remove this logic and move on.

`**test_views.py**`

- `ConfigContextTestCase` - `#` `*TODO: Change base class to PrimaryObjectViewTestCase*`
    - Blocked by absence of standard create/edit, bulk create views
- `ObjectChangeTestCase` - `# TODO: Convert to StandardTestCases.Views`
    - Self-explantory
- `RelationshipAssociationTestCase` - Do we need to support create/update/retrieve for `RelationshipAssociation`? See:
    - `*# TODO? ViewTestCases.CreateObjectViewTestCase,*`
    - `#` `*TODO? ViewTestCases.EditObjectViewTestCase,*`
    - `#` `*TODO? ViewTestCases.GetObjectViewTestCase,*`

## `views.py`

`**ConfigContext*View**`

This appears to be for all `ConfigContext` and `ConfigContextSchema` update views?

```
    # TODO(Glenn): disallow (or at least warn) user from manually editing config contexts that
    # have an associated owner, such as a Git repository
```

and

```
    # TODO: disallow (or at least warn) user from manually editing config context schemas that
    # have an associated owner, such as a Git repository
```

- Needs clarification

`**DynamicGroupEditView**`

- `get_extra_context`

```
    # FIXME(jathan): There is currently an edge case here that needs to be addressed:
    # `AttributeError: 'NoneType' object has no attribute 'is_valid'`
    # See: https://sentry.io/share/issue/fb41c6afb40248f6931021574bc38a0d/
```

- This may no longer be an edge case and instead may be an expected scenario. Need to confirm.

# IPAM

## `tests`

`**test_views.py::ServiceTestCase**`

```
    # TODO: Update base class to PrimaryObjectViewTestCase
```

- Blocked by absence of standard creation view

## `views.py`

`PrefixListView.queryset`

```
    TODO(john): When the base views support a formal `get_queryset()` method, this approach is not needed
```

- This should be obviated as we move to `NautobotViewSet` patterns for core apps

# Users

## `api/views.py`

`**UserConfigViewSet.patch**`

```
    # TODO: How can we validate this data?
```

How can we validate this data? A serializer, of course.

## `views.py`

`**InterfaceConnectionsListView.get_queryset**`

```
    # TODO: this is very problematic when filtering the view via FilterSet - if the filterset matches (A), then
    #       the connection will appear in the table, but if it only matches (B) then the connection will not!
```

This IS problematic and it seems like a tough one to solve.

# Utilities

> This entire app will go away in v2, but the primitives will likely remain in a new location.

## `filters.py`

`**NaturalKeyOrPKMultipleChoiceFilter.get_filter_predicate**`

```
    # FIXME(jathan): It feels weird to have a list/qs make its way to `get_filter_predicate()`
    # which if you inspect the source for `django_filters.MultipleChoiceFilter.filter()` should
    # be handling those to generate singular filter predicates and concatentating them.
```

- Let’s see if we can remove the need for this code or just remove the FIXME.

## `forms/fields.py`

`**CommentField**`

```
    # TODO: Port Markdown cheat sheet to internal documentation
```

- I am not convinced this is necessary, although it was something that came to fruition in NetBox (see: https://docs.netbox.dev/en/stable/reference/markdown/), I still don’t think we have to embed this directly.

`**DynamicModelChoiceMixin**`

```
    # TODO(Glenn): can we change this? pylint: disable=redefined-builtin
```

- It’s not clear on why woudl want to chance this? Is it because it is assigning to the name `filter` which `pylint` things is defining the `filter()` built-in?
    - For example, changing this to be a property or an instance variable?

`**JSONField.bound_data**`

```
    # TODO: remove this when we upgrade to Django 4
```

- Why tho? Is there even any need to overload `JSONField` now that it’s a built-in Django field?
- Ask Dan why he added this comment.

## `query_functions.py`

`**JSONBAgg**`

```
    # TODO(Glenn): Django's JSONBAgg has `allow_distinct=True`, we might want to think about adding that at some point?
```

- Why tho?

## `testing`

`**__init__.py**`

- `CeleryTestCase.wait_on_active_tasks`

```
    # TODO(john): admittedly, this is not great, but it seems the standard
    # celery APIs for inspecting the worker, looping through all active tasks,
    # and calling `.get()` on them is not working when the worker is in solo mode.
    # Needs more investigation and until then, these tasks run very quickly, so
    # simply delaying the test execution provides enough time for them to complete.
```

    - Having to sleep 1 second sucks. It would be nice if we didn’t have to. It’s not clear on our path forward here without a time investment to arrive at a solution.

`**api.py**`

- `ListObjectsViewTestCase.test_list_objects_anonymous`

```
    # TODO FIXME: if we're passing **self.header, we are *by definition* **NOT** anonymous!!
```

    - We should eliminate the need to pass `self.header` on every call and just make it implicit.
    - All the same this comment isn’t clear. What is in `self.header` here that is belying anonymity?

- `UpdateObjectViewTestCase.test_bulk_update_objects`

```
    # TODO(Glenn): shouldn't we also check that obj[field] == value?
```

- Yes. So let’s do that.

`integration.py`

- `SeleniumTestCase.requires_celery`

```
    # If true, a celery instance will be started. TODO: create celery mixin?
```

    - We have a `CeleryTestCase` and a `NautobotTestCaseMixin`
    - But in either case, setting a class attribute or including a mixin are somewhat equivalent in being declarative. We should think about if we want composition to be a requirement here.

- `SeleniumTestCase.wait_on_active_tasks`

```
    # TODO(john): admittedly, this is not great, but it seems the standard
    # celery APIs for inspecting the worker, looping through all active tasks,
    # and calling `.get()` on them is not working when the worker is in solo mode.
    # Needs more investigation and until then, these tasks run very quickly, so
    # simply delaying the test execution provides enough time for them to complete.
```

    - There is the compelling argument for a Celery mixin. We have code duplication between `CeleryTestCase`  and `SeleniumTestCase`.

## `views.py`

`**ModelViewTestCase**`

- `_get_base_url`
- `_get_url`

```
    # 2.0 TODO(jathan): Eliminate the need to overload `_get_base_url()` at all and just rely on `get_route_for_model()`
```

    and

```
    # 2.0 TODO(jathan): Eliminate the need to overload `_get_url()` at all and just rely on `get_route_for_model()`
```

    - The need to overload these are the exception and not the norm

`**ListObjectsViewTestCase**`

- `test_list_objects_filtered`
- `test_list_objects_unknown_filter_strict_filtering`
- `test_list_objects_unknown_filter_no_strict_filtering`
- `test_list_objects_with_constrained_permission`

```
    # TODO: it'd make test failures more readable if we strip the page headers/footers from the content
```

    - More assertion that we should make `extract_page_body` strip headers/footers

## `tests/test_templatetags_helpers.py`

`**NautobotTemplatetagsHelperTest**`

- `test_tzoffset` - `#` `*TODO add unit tests for tzoffset*`
- `test_has_perms` - `#` `*TODO add unit tests for has_perms*`
- `test_has_one_or_more_perms` - `#` `*TODO add more unit tests for has_one_or_more_perms*`

# Virtualization

## `forms.py`

`ClusterAddDevicesForm.location`

```
    # TODO: base_site: $site
```

- Why tho? Can we get rid of this?

## `views.py`

# 2.0 TODOs

Things here should also be backlogged.

But this needs to be a secondary effort.

See: https://github.com/nautobot/nautobot/issues/2582

**Comment by glennmatthews** [comment]
> ## `pyproject.toml`
> 
> There is a rather large list of disabled rules for PyLint that we should revisit.

I did some analysis of these in https://github.com/nautobot/nautobot/issues/2292#issuecomment-1235666827


> `**nautobot.utilities.testing.views.APIViewTestCase**`
> 
> ```
>     # TODO: Standardize to APIViewTestCase (needs create & update tests)
> ```
> 
>     * Self-explanatory in that create/update generic tests are needed?

Yes - if we can add create/update API endpoints for `ImageAttachment`, then `ImageAttachmentTest` can use the full-featured `APIViewTestCase`.


> `**ObjectEditViewMixin**`
> 
> Each of the following methods are overloading methods to work with forms vs. serializers. We should revisit as we move towards serializers as forms.
> 
>     * `perform_create` → `#` `*TODO: this conflicts with DRF's CreateModelMixin.perform_create(self, serializer) API*`
> 
>     * `perform_update` → `#` `*TODO: this conflicts with DRF's UpdateModelMixin.perform_update(self, serializer) API*`
> 
> 
> `**ObjectBulkUpdateViewMixin**`
> 
> We should revisit as we move towards serializers as forms.
> 
>     * `perform_bulk_update` → `#` `*TODO: this conflicts with BulkUpdateModelMixin.perform_bulk_update(self, objects, update_data, partial)*`

Changing these is *probably* a breaking (i.e. 2.0) change since plugins using the NautobotUIViewSet may be overriding these functions locally.



> ## `constants.py`
> 
> `JOB_MAX_NAME_LENGTH`
> 
> ```
>     # TODO(Glenn): this should really be a more general-purpose constant, like NAME_MAX_LENGTH
> ```
> 
>     * If so, is this something we would use in other places? What are the references?

I.e., we have SO MANY models that are explicitly defining `name = models.CharField(max_length=100)`, but if we're committed to self-consistency it should be a global constant instead.

> 
> ```
>     # FIXME: Normally ObjectChange records are automatically generated every time we save an object,
>     # regardless of whether any fields were actually modified.
>     # Because a single GitRepository may manage dozens of records, this would result in a lot of noise
>     # every time a repository gets resynced.
>     # To reduce that noise until the base issue is fixed, we need to explicitly detect object changes:
> ```
> 
>     * It’s not clear what the “base issue” here is? That `ObjectChange` objects are generated every time we save and not only when a change occurs?

Yes.

> 
> 
> `import_local_config_context`
> 
> ```
>     # TODO: come up with a design that accounts for non-unique names, as well as un-named Devices.
> ```
> 
>     * This is because `Device.name` is not guaranteed globally unique
> 
>     * The correct solution is to make `Device.name` globally unique in a future release.
>       
>       * Further un-named devices should not be possible. Sorry.
>       * Object in inventory without hostnames should still have hostnames.

How to address the case of "This is a piece of hardware sitting on a shelf that hasn't yet been installed" then? What should the name of such a device be?




> 
> ```
>     # TODO(Glenn): catch DB error in case where multiple Jobs have the same grouping + name
> ```
> 
>     * IMO It should not be possible for multiple jobs to have the same grouping + name

The problem is that both `grouping` and `name` can be defined in the Job class `Meta` data - so we are responsible for detecting and handling at import/discovery time any cases where for whatever reason two different Jobs declare identical values.

> `**JobTable**`
> 
> ```
>     # TODO(Glenn): pk = ToggleColumn()
> ```
> 
>     * Why isn’t this already the case?

I think I just didn't get to it when I did the Jobs work for 1.3.

> `**test_datasources.py**`
> 
>     * `GitTest.populate_repo`  - `#` `*TODO: populate Jobs as well?*`
>       
>       * This seems to just be populating test `jobs` published by a Git repo?

This is our mock Git repository contents (so that we don't need an actual external Git repository to sync to). Right now the mock repository doesn't contain any Job files, so our testing for Git repository behavior doesn't include the Job loading logic. This is a gap.

> ## `views.py`
> 
> `**ConfigContext*View**`
> 
> This appears to be for all `ConfigContext` and `ConfigContextSchema` update views?
> 
> ```
>     # TODO(Glenn): disallow (or at least warn) user from manually editing config contexts that
>     # have an associated owner, such as a Git repository
> ```
> 
> and
> 
> ```
>     # TODO: disallow (or at least warn) user from manually editing config context schemas that
>     # have an associated owner, such as a Git repository
> ```
> 
>     * Needs clarification

This is part of the broader SSOT "data ownership" feature request (which actually, I'm not sure we have an issue tracking in the core repo at this time). Basically we want to be able to mark any data as "externally owned", i.e. it exists in Nautobot but Nautobot is not the system-of-record for it, so Nautobot's replica of the data should be read-only to most users to avoid confusion and introducing mismatched data.

For ConfigContext and ConfigContextSchema (and ExportTemplate, IIRC), we actually already have an "owner" FK to GitRepository that gets set when a GitRepository is responsible for creating/updating specific instances of these models. However, the presence of that FK does not currently prevent a user from manually editing those instances (their manual edits will be overwritten next time the repository gets synced though) -- this TODO is about changing that behavior / closing that gap.


> `**DynamicModelChoiceMixin**`
> 
> ```
>     # TODO(Glenn): can we change this? pylint: disable=redefined-builtin
> ```
> 
>     * It’s not clear on why woudl want to chance this? Is it because it is assigning to the name `filter` which `pylint` things is defining the `filter()` built-in?
>       
>       * For example, changing this to be a property or an instance variable?

Just renaming it to something like `_filter` or `filter_` to make pylint happy, yes. I think it's not an inherent property of any base class (in which case renaming it would probably be problematic), but specific to this mixin class, so we should be able to do so safely.

> `**JSONBAgg**`
> 
> ```
>     # TODO(Glenn): Django's JSONBAgg has `allow_distinct=True`, we might want to think about adding that at some point?
> ```
> 
>     * Why tho?

Because our JSONBAgg has different defaults than Django's own JSONBAgg class - I'm not sure what the implications of that delta are.

> `**api.py**`
> 
>     * `ListObjectsViewTestCase.test_list_objects_anonymous`
> 
> 
> ```
>     # TODO FIXME: if we're passing **self.header, we are *by definition* **NOT** anonymous!!
> ```
> 
> ```
> - We should eliminate the need to pass `self.header` on every call and just make it implicit.
> - All the same this comment isn’t clear. What is in `self.header` here that is belying anonymity?
> ```

See `setUp`:
`        self.header = {"HTTP_AUTHORIZATION": f"Token {self.token.key}"}`

So if we're passing `self.header` in our API call, we're providing a user token in the API call, so we're not making an anonymous request.

> # Virtualization
> ## `forms.py`
> 
> `ClusterAddDevicesForm.location`
> 
> ```
>     # TODO: base_site: $site
> ```
> 
>     * Why tho? Can we get rid of this?

#2518

**Comment by jathanism** [comment]
> How to address the case of "This is a piece of hardware sitting on a shelf that hasn't yet been installed" then? What should the name of such a device be?

Inventory items? Asset hostnames that are immutable (like a slug but unique to that device)? We have the technology! There needs to be some kind of natural key for devices that doesn't rely on duplicates or null values.

**Comment by jathanism** [comment]
> The problem is that both grouping and name can be defined in the Job class Meta data - so we are responsible for detecting and handling at import/discovery time any cases where for whatever reason two different Jobs declare identical values.

That clarification helps. More like detect and error, not detect and allow? I'm good with that.

**Comment by glennmatthews** [comment]
> > How to address the case of "This is a piece of hardware sitting on a shelf that hasn't yet been installed" then? What should the name of such a device be?
> 
> Inventory items? Asset hostnames that are immutable (like a slug but unique to that device)? We have the technology! There needs to be some kind of natural key for devices that doesn't rely on duplicates or null values.

I think we've talked about maybe making it so that a device *must* have at minimum either a name or a serial number. That to my mind would be sufficient.
